helix-core/src/main/java/org/apache/helix/model/CloudConfig.java

```diff
@@ -0,0 +1,289 @@
+          _cloudInfoSources, _cloudInfoProcessorName);
+    }
+
+    /**
+     * Default constructor
+     */
+    public Builder() {
+    }
+
+    /**
+     * Constructor with Cluster Name as input
+     * @param clusterName
+     */
+    public Builder(String clusterName) {
+      _clusterName = clusterName;
+    }
+
+    /**
+     * Constructor with CloudConfig as input
+     * @param cloudConfig
+     */
+    public Builder(CloudConfig cloudConfig) {
+      _cloudEnabled = cloudConfig.isCloudEnabled();
+      _cloudProvider = CloudProvider.valueOf(cloudConfig.getCloudProvider());
+      _cloudID = cloudConfig.getCloudID();
+      _cloudInfoSources = cloudConfig.getCloudInfoSources();
+      _cloudInfoProcessorName = cloudConfig.getCloudInfoProcessorName();
+    }
+
+    public Builder setClusterName(String v) {
+      _clusterName = v;
+      return this;
+    }
+
+    public Builder setCloudEnabled(boolean isEnabled) {
+      _cloudEnabled = isEnabled;
+      return this;
+    }
+
+    public Builder setCloudProvider(CloudProvider cloudProvider) {
+      _cloudProvider = cloudProvider;
+      return this;
+    }
+
+    public Builder setCloudID(String v) {
+      _cloudID = v;
+      return this;
+    }
+
+    public Builder setCloudInfoSources(List<String> v) {
+      _cloudInfoSources = v;
+      return this;
+    }
+
+    public Builder addCloudInfoSource(String v) {
+      if (_cloudInfoSources == null) {
+        _cloudInfoSources = new ArrayList<String>();
+      }
+      _cloudInfoSources.add(v);
+      return this;
+    }
+
+    public Builder setCloudInfoProcessorName(String v) {
+      _cloudInfoProcessorName = v;
+      return this;
+    }
+
+    public String getClusterName() {
+      return _clusterName;
+    }
+
+    public CloudProvider getCloudProvider() {
+      return _cloudProvider;
+    }
+
+    public boolean getCloudEnabled() {
+      return _cloudEnabled;
+    }
+
+    public String getCloudID() {
+      return _cloudID;
+    }
+
+    public List<String> getCloudInfoSources() {
+      return _cloudInfoSources;
+    }
+
+    public String getCloudInfoProcessorName() {
+      return _cloudInfoProcessorName;
+    }
+
+    private void validate() {
+      if (_cloudEnabled) {
+        if (_cloudID == null) {
+          throw new HelixException(
+              "This Cloud Configuration is Invalid. The CloudID is missing from the config.");
+        }
+        if (_cloudProvider == null) {
+          throw new HelixException(
+              "This Cloud Configuration is Invalid. The Cloud Provider is missing from the config.");
+        } else if (_cloudProvider == CloudProvider.CUSTOM) {
+          if (_cloudInfoProcessorName == null) {
+            throw new HelixException(
+                "This Cloud Configuration is Invalid. Custom provider has been chosen without defining CloudInfoProcessorName.");
+          }
+          if (_cloudInfoSources == null || _cloudInfoSources.size() == 0) {
```

Comment:
Combing these two. We just need let user know either processor or source is invalid.

Reply:
Done.